File: backend/app/routers/stream.py
import asyncio
import json

from fastapi import APIRouter
from fastapi.responses import StreamingResponse

from ..core.database import get_db

router = APIRouter(prefix="/stream", tags=["stream"])


@router.get("/events")
async def sse_stream():
    """
    Server-Sent Events endpoint.

    Polls MongoDB for changes every 3 seconds and pushes update
    notifications to connected clients.
    """

    async def event_generator():
        last_check = ""
        while True:
            await asyncio.sleep(3)
            try:
                db = get_db()
                latest = await db["people"].find_one(
                    {"last_seen": {"$ne": None}},
                    {"last_seen": 1},
                    sort=[("last_seen", -1)],
                )
                current = latest["last_seen"] if latest else ""
                if current != last_check:
                    last_check = current
                    payload = json.dumps({"type": "update", "ts": current})
                    yield f"data: {payload}\n\n"
            except Exception:
                pass

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )

File: src/server-routes/stream.ts
import { Router } from "express";
import { getDb } from "../server-core/database";
import { authMiddleware } from "../server-core/security";
import { resolvePatientId } from "../server-core/patientResolver";

const router = Router();

// GET /stream/events (SSE)
router.get("/events", authMiddleware, resolvePatientId, (req, res) => {
  res.setHeader("Content-Type", "text/event-stream");
  res.setHeader("Cache-Control", "no-cache");
  res.setHeader("X-Accel-Buffering", "no");
  res.setHeader("Connection", "keep-alive");
  res.flushHeaders();

  let lastCheck = "";
  const interval = setInterval(async () => {
    try {
      const db = getDb();
      const latest = await db
        .collection("people")
        .findOne(
          { patient_id: req.patientId!, last_seen: { $ne: null } },
          { projection: { last_seen: 1 }, sort: { last_seen: -1 } }
        );

      const current = latest?.last_seen || "";
      if (current !== lastCheck) {
        lastCheck = current;
        res.write(`data: ${JSON.stringify({ type: "update", ts: current })}\n\n`);
      }
    } catch {}
  }, 3000);

  req.on("close", () => {
    clearInterval(interval);
  });
});

export default router;
